src/shims/backtrace.rs

```diff
@@ -49,31 +69,87 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             })
             .collect();
 
-        let len = ptrs.len();
+        let len = ptrs.len().try_into().unwrap();
 
         let ptr_ty = tcx.mk_ptr(TypeAndMut { ty: tcx.types.unit, mutbl: Mutability::Mut });
 
-        let array_ty = tcx.mk_array(ptr_ty, ptrs.len().try_into().unwrap());
+        let array_ty = this.layout_of(tcx.mk_array(ptr_ty, len)).unwrap();
 
-        // Write pointers into array
-        let alloc =
-            this.allocate(this.layout_of(array_ty).unwrap(), MiriMemoryKind::Rust.into())?;
-        for (i, ptr) in ptrs.into_iter().enumerate() {
-            let place = this.mplace_index(&alloc, i as u64)?;
-            this.write_pointer(ptr, &place.into())?;
-        }
+        match flags {
+            // storage for pointers is allocated by miri
+            // deallocating the slice is undefined behavior with a custom global allocator
+            0 => {
+                let &[_] = this.check_shim(abi, Abi::Rust, link_name, args)?;
+
+                let alloc = this.allocate(array_ty, MiriMemoryKind::Rust.into())?;
+
+                // Write pointers into array
+                for (i, ptr) in ptrs.into_iter().enumerate() {
+                    let place = this.mplace_index(&alloc, i as u64)?;
+
+                    this.write_pointer(ptr, &place.into())?;
+                }
+
+                this.write_immediate(
+                    Immediate::new_slice(Scalar::from_maybe_pointer(alloc.ptr, this), len, this),
+                    dest,
+                )?;
+            }
+            // storage for pointers is allocated by the caller
+            1 => {
+                let &[_, ref buf] = this.check_shim(abi, Abi::Rust, link_name, args)?;
+
+                this.memory.check_ptr_access_align(
+                    this.read_pointer(buf)?,
+                    array_ty.size,
+                    array_ty.align.abi,
+                    CheckInAllocMsg::MemoryAccessTest,
+                )?;
```

Comment:
I don't think this is needed, the methods called below should already do the required checks.